aten/src/ATen/core/ivalue.cpp

```diff
@@ -604,4 +604,55 @@ getClassConverter() {
       classConverter;
   return classConverter;
 }
+
+CAFFE2_API intrusive_ptr<ivalue::Future> collectAll(
+    List<intrusive_ptr<ivalue::Future>> srcs) {
+  struct Ctx {
+    explicit Ctx(List<intrusive_ptr<ivalue::Future>> srcs)
+        : remaining(srcs.size()),
+          srcFutures(std::move(srcs)),
+          asIvalue(srcFutures),
+          dstFuture(make_intrusive<ivalue::Future>(asIvalue.type())) {}
+    std::atomic<int32_t> remaining{0};
+    List<intrusive_ptr<ivalue::Future>> srcFutures;
+    IValue asIvalue;
+    intrusive_ptr<ivalue::Future> dstFuture;
+  };
+
+  auto ctx = std::make_shared<Ctx>(std::move(srcs));
+  std::function<void()> func = [ctx]() {
+    if (--ctx->remaining == 0) {
+      ctx->dstFuture->markCompleted(ctx->asIvalue);
+    }
+  };
+  for (int32_t tot = ctx->srcFutures.size(), i = 0; i < tot; ++i) {
+    ctx->srcFutures.get(i)->addCallback(func);
+  }
+  if (ctx->srcFutures.size() == 0) {
+    ctx->dstFuture->markCompleted(ctx->asIvalue);
+  }
+  return ctx->dstFuture;
+}
+
+CAFFE2_API intrusive_ptr<ivalue::Future> collectAll(
+    std::vector<intrusive_ptr<ivalue::Future>> srcs) {
+  auto consistentTypePtr =
+      [](const std::vector<intrusive_ptr<ivalue::Future>>& srcs) -> TypePtr {
+    if (srcs.empty()) {
+      return FutureType::create(AnyType::get());
+    }
+    TypePtr res = srcs[0]->elementType();
+    for (size_t i = 1, len = srcs.size(); i < len; ++i) {
+      if (*srcs[i]->elementType() != *res) {
+        return FutureType::create(AnyType::get());
```

Comment:
I guess if there's no use case that takes multiple future of different types into a single list, we should error out to the user in this case rather than use the AnyType?

Reply:
There is no current use case, though if it's simple to enable, not clear why it needs to be ane error?

Reply:
I feel like this will break the list type constraint, e.g. a list like `List{Future[Int], Future[Double]}` is wrong in theory. For a single list instance, we should always have the same `element_type` for all the elements, so user have to construct two list with Future[Int] and Future[Double] separately and call `collectAll` separately, but we seems enabling this wrong pattern in this PR.

Reply:
That's fine, I'll add a `TORCH_CHECK()`